### ISSUE TITLE
feat: support named custom formats in terraform_naming_convention rule

### ISSUE THREAD

**Issue by angelyan** [terraform-linters/tflint#857]
When a custom format check fails, the error message just shows the regular expression:

```
module name `my_module` must match the following RegExp: ^[A-Z][a-z]+(?:[A-Z][a-z]+)*$
```

It would be nice if you could set names for these custom expressions and show _them_ instead, the same way the predefined formats are shown:

```
module name `my_module` must match the following format: pascal_case
```

As for the configuration interface, an option would be to add an attribute (e.g. `custom_formats`) consisting of a map of names to expressions. These names could then be used in the `format` attribute. For example:

```hcl
rule "terraform_naming_convention" {
  enabled = true
  custom_formats = {
    camel_case  = "^[a-z]+(?:[A-Z][a-z]+)*$"
    pascal_case = "^[A-Z][a-z]+(?:[A-Z][a-z]+)*$"
  }

  format = "camel_case"

  module {
    format = "pascal_case"
  }
} 
```

If you think this is a good idea, I could start working on a pull request.

**Comment by wata727** [comment]
In the case (`pascal_case`), it would be better to add the format to predefined formats. Initially, more formats were planned to be added, but omitted for simplicity. https://github.com/terraform-linters/tflint/pull/697#issuecomment-609041137

This is useful for cases where you want to name a special pattern that we can't add to predefined formats, but I'd like to think carefully about whether there is such a case.

**Comment by bendrucker** [comment]
I'd actually like to see snake_case treated as the only built in format, with everything else being user supplied. The standard block label case is snake case. 

It sort of surprises me to see demand for this level of complexity. It's not trivial to rename, but following up with `terraform state mv`, particularly if it's just modules, is always an option. 

Shoe-horning descriptive info into the keys seems like a mistake, perhaps more like:

```hcl
custom_formats = {
  camel_case   = {
    description = "camelCase"
    regex       = "^[a-z]+(?:[A-Z][a-z]+)*$"
  }
}
```

The more specific your format the harder it'll be to describe, and a non-identifier gives you the freedom to mimic the desired case, add a parenthetical, use spaces, etc. I can dream up a bunch of contrived/complex formats (digits are valid too, for example), but this feature is accommodating non-standard formatting in the wild, so it's hard to say just how contrived they would be.

**Comment by angelyan** [comment]
I like your suggestion; using the key as a description is not the best idea.

I believe that providing a descriptive message is almost always better than just showing a regular expression. Since custom expressions are allowed, I don't see why supporting such descriptions would be undesirable.

**Comment by bendrucker** [comment]
Great, happy to review a PR. Seems like a breaking change here is inevitable in order to ultimately better serve the end users who are met with these lint failures. It'll only affect those using custom formats, so hopefully a smallish slice of users of this rule.

**Comment by angelyan** [comment]
Custom formats could be kept if backwards compatibility is desired, unless you prefer for me to remove them.

**Comment by bendrucker** [comment]
True, I guess we can look at dropping separately. Though part of the reason TFLint is an 0.x product is for frequent breaks like this.
